### PR TITLE
Refactoring publisher.go to use the new publish_* scripts

### PR DESCRIPTION
**Pull request by caesarxuchao**
The first commit greatly simplified publisher.go.
Second commit is for my local debugging.

I removed the issue filer from this PR. I'll introduce it back after this one gets merged.

My local runs are successful. The generated changes are pushed to:
https://github.com/kubernetes/apimachinery/pull/9
https://github.com/kubernetes/client-go/pull/141
https://github.com/kubernetes/client-go/pull/140

I checkout the generated master branches locally, and run "go build ./..." in client-go, it's successful. Note that client-go doesn't have apimachinery in its vendor, so this means client-go's master works with apimachinery's master.

I suggest that we proceed this way:
1. we first review the three generated pulls; 
2. merge the apimachinery one first; that should make travis happy in https://github.com/kubernetes/client-go/pull/140.
3. merge the two generated client-go pulls.
4. merge this pull and turn on the robot.

### CODE REVIEW

**Comment by fejta** [comment]
/unassign @stts @grodrigues3 
NM just noticed the WIP field

File: mungegithub/mungers/publish_scripts/push.sh [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+set -o pipefail
+
+if [ ! $# -eq 2 ]; then
+    echo "usage: $0 token branch"
+    exit 1
+fi
+
+TOKEN="${1}"
+BRANCH="${2}"
+readonly TOKEN
```
Comment (sttts) [inline mungegithub/mungers/publish_scripts/push.sh]: nit: BRANCH

File: mungegithub/mungers/publisher.go [hunk 2]
```diff
@@ -70,14 +75,14 @@ func init() {
...
-func (p *PublisherMunger) RequiredFeatures() []string { return []string{features.RepoFeatureName} }
+func (p *PublisherMunger) RequiredFeatures() []string { return []string{} }
 
 // Initialize will initialize the munger
 func (p *PublisherMunger) Initialize(config *github.Config, features *features.Features) error {
-	p.baseDir = features.Repos.BaseDir
-	if len(p.baseDir) == 0 {
-		glog.Fatalf("--repo-dir is required with selected munger(s)")
-	}
+	// the $PATH set in Dockerfile will be overwritten by /etc/profile, so we have to set the PATH again here.
```
Comment (sttts) [inline mungegithub/mungers/publisher.go]: where is this done?

File: mungegithub/mungers/publisher.go [hunk 3]
```diff
@@ -111,133 +128,80 @@ func clone(dst string, dstURL string) error {
...
-}
-
-// EachLoop is called at the start of every munge loop
-func (p *PublisherMunger) EachLoop() error {
-	var errlist []error
-Repos:
-	for _, rules := range p.reposRules {
+// constructs all the repos, but does not push the changes to remotes.
+func (p *PublisherMunger) construct() error {
+	for _, repoRules := range p.reposRules {
```
Comment (sttts) [inline mungegithub/mungers/publisher.go]: s/, repoRules/, repoRule/

File: mungegithub/mungers/publisher.go [hunk 4]
```diff
@@ -111,133 +128,80 @@ func clone(dst string, dstURL string) error {
...
+	// NOTE: because some repos depend on each other, e.g., client-go depends on
+	// apimachinery, they should be published atomically, but it's not supported
+	// by github.
+	for _, repoRules := range p.reposRules {
+		dstDir := filepath.Join(p.kubernetesPath, repoRules.dstRepo, "")
+		if err := os.Chdir(dstDir); err != nil {
+			return err
+		}
+		for src, dst := range repoRules.srcToDst {
+			cmd := exec.Command("/publish_scripts/push.sh", src.branch, dst.branch)
```
Comment (sttts) [inline mungegithub/mungers/publisher.go]: shouldn't the first param be the token?

**Comment by sttts** [comment]
Looks much better now 👍

File: mungegithub/mungers/publisher.go (hunk 2 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publisher.go]: The comment is removed, it was wrong.

File: mungegithub/mungers/publisher.go (hunk 3 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publisher.go]: a single repo may have multiple branches, so it's "repoRules"

File: mungegithub/mungers/publisher.go (hunk 4 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publisher.go]: Fixed.

**Comment by caesarxuchao** [comment]
@sttts this is ready now. Could we proceed as i suggested at the end of the first commit?

**Comment by sttts** [comment]
@caesarxuchao you mean as in the 4 steps in the PR description. I think we should. Do I owe you any review anywhere? Do you need any help?

**Comment by caesarxuchao** [comment]
I need some help reviewing the 3 generated PRs. Could you or @deads2k review the apimachinery [one](https://github.com/kubernetes/apimachinery/pull/9)? Me and @mbohlool will take another look at the two client-go PRs. 
Please let us know if you merge the apimachinery one, we'll then merge the client-go ones so that they are synced.

**Comment by caesarxuchao** [comment]
@sttts have you reviewed the existing code? If so, can we merge the current code as is? I sent a WIP https://github.com/kubernetes/test-infra/pull/2272 to let the publish robot work for more repos, it actually changes the scripts a little bit to reuse more code, so might be easier to review as a standalone pr.

File: mungegithub/Dockerfile-publisher [hunk 5]
```diff
@@ -15,9 +15,20 @@
...
+RUN apt-get install -y -qq ca-certificates git wget jq
+RUN wget https://storage.googleapis.com/golang/go1.7.5.linux-amd64.tar.gz
+RUN tar -C /usr/local -xzf go1.7.5.linux-amd64.tar.gz
+ENV PATH="/usr/local/go/bin:${PATH}"
+
+ENV GOPATH="/go-workspace"
+ENV PATH="${GOPATH}/bin:${PATH}"
+RUN go get github.com/tools/godep
+WORKDIR "/go-workspace/src/github.com/tools/godep"
+RUN git checkout tags/v75
```
Comment (sttts) [inline mungegithub/Dockerfile-publisher]: v79 will fix issues with Go 1.8. I wouldn't mind to bump this here as well, as in https://github.com/kubernetes/kubernetes/pull/42967.

File: mungegithub/mungers/publisher.go [hunk 6]
```diff
@@ -70,14 +75,13 @@ func init() {
...
+func (p *PublisherMunger) RequiredFeatures() []string { return []string{} }
 
 // Initialize will initialize the munger
 func (p *PublisherMunger) Initialize(config *github.Config, features *features.Features) error {
-	p.baseDir = features.Repos.BaseDir
-	if len(p.baseDir) == 0 {
-		glog.Fatalf("--repo-dir is required with selected munger(s)")
-	}
+	gopath := os.Getenv("GOPATH")
+	p.kubernetesPath = filepath.Join(gopath, "src", "k8s.io")
```
Comment (sttts) [inline mungegithub/mungers/publisher.go]: k8sIOPath

**Comment by sttts** [comment]
@caesarxuchao code here lgtm (commented some nits, feel free to fix them or ignore them).

**Comment by caesarxuchao** [comment]
Thanks @sttts. I'll address the two new comments in the followup. @krzyzacy could you help merging this PR, i got the lgtm [here](https://github.com/kubernetes/test-infra/pull/2216#issuecomment-287003426)?

**Comment by krzyzacy** [comment]
@caesarxuchao I pulled you into test-infra-maintainers, you should have merge now here.

**Comment by caesarxuchao** [comment]
Cool. Thanks. Merging.

**Comment by sttts** [comment]
:rocket:

Is this active from now on, i.e. publishing daily?

**Comment by caesarxuchao** [comment]
Not yet. We need to get those manual PRs merged first :(

**Comment by sttts** [comment]
@caesarxuchao let me know if you need a review or any other help on the manual PRs.

**Comment by caesarxuchao** [comment]
The bot has some issues updating the k8s.io/apiserver repo. I'll ping you when it's ready. Thanks!

File: mungegithub/Dockerfile-publisher (hunk 5 above)
Comment (caesarxuchao) [inline mungegithub/Dockerfile-publisher]: Done.

File: mungegithub/mungers/publisher.go (hunk 6 above)
Comment (caesarxuchao) [inline mungegithub/mungers/publisher.go]: Done.

**Pull request merged**